File: src/worker/views/account.rb
# Translated from TypeScript to Ruby
# This code uses the Ruby standard library.

GRACE_PERIOD = 60 * 60 * 24 * 3

def update_account(inputs, update_events, state, conditions)
  if update_events.include?("firstRun") || update_events.include?("account")
    partial_top_menu = check_account(conditions)
    logged_in = !partial_top_menu[:username].nil? && !partial_top_menu[:username].empty?
    gold_until_date = Time.at(partial_top_menu[:goldUntil])
    gold_until_date_string = gold_until_date.strftime("%a %b %d %Y")
    current_timestamp = (Time.now.to_i - GRACE_PERIOD)
    show_gold_active = logged_in && !partial_top_menu[:goldCancelled] && current_timestamp < partial_top_menu[:goldUntil]
    show_gold_cancelled = logged_in && partial_top_menu[:goldCancelled] && current_timestamp < partial_top_menu[:goldUntil]
    show_gold_pitch = !logged_in || !show_gold_active

    return {
      email: partial_top_menu[:email],
      gold_message: inputs[:goldMessage],
      gold_success: inputs[:goldSuccess],
      gold_until_date_string: gold_until_date_string,
      logged_in: logged_in,
      show_gold_active: show_gold_active,
      show_gold_cancelled: show_gold_cancelled,
      show_gold_pitch: show_gold_pitch,
      username: partial_top_menu[:username]
    }
  end
end

File: src/worker/views/account.hs
-- Framework: Haskell

import Data.Time.Clock.POSIX (getPOSIXTime)
import qualified Data.List as List

-- For subscribers who have not renewed yet, give them a 3 day grace period before showing ads again, because sometimes it takes a little extra time for the payment to process
gracePeriod :: Int
gracePeriod = 60 * 60 * 24 * 3

updateAccount :: ViewInput -> [String] -> () -> Conditions -> IO (Maybe AccountUpdate)
updateAccount inputs updateEvents state conditions = do
    currentTimestamp <- fmap floor getPOSIXTime
    let adjustedCurrentTimestamp = currentTimestamp - gracePeriod
    
    if "firstRun" `elem` updateEvents || "account" `elem` updateEvents
        then do
            partialTopMenu <- checkAccount conditions
            let loggedIn = not (null (username partialTopMenu))
                goldUntilDate = posixSecondsToUTCTime (fromIntegral (goldUntil partialTopMenu))
                goldUntilDateString = show goldUntilDate
                showGoldActive = loggedIn && not (goldCancelled partialTopMenu) && adjustedCurrentTimestamp < goldUntil partialTopMenu
                showGoldCancelled = loggedIn && goldCancelled partialTopMenu && adjustedCurrentTimestamp < goldUntil partialTopMenu
                showGoldPitch = not loggedIn || not showGoldActive
            
            return $ Just AccountUpdate {
                email = email partialTopMenu,
                goldMessage = goldMessage inputs,
                goldSuccess = goldSuccess inputs,
                goldUntilDateString = goldUntilDateString,
                loggedIn = loggedIn,
                showGoldActive = showGoldActive,
                showGoldCancelled = showGoldCancelled,
                showGoldPitch = showGoldPitch,
                username = username partialTopMenu
            }
        else
            return Nothing

data ViewInput = ViewInput {
    goldMessage :: String,
    goldSuccess :: Bool
}

data Conditions = Conditions -- Define conditions structure here

data PartialTopMenu = PartialTopMenu {
    username :: String,
    email :: String,
    goldUntil :: Int,
    goldCancelled :: Bool
}

data AccountUpdate = AccountUpdate {
    email :: String,
    goldMessage :: String,
    goldSuccess :: Bool,
    goldUntilDateString :: String,
    loggedIn :: Bool,
    showGoldActive :: Bool,
    showGoldCancelled :: Bool,
    showGoldPitch :: Bool,
    username :: String
}

checkAccount :: Conditions -> IO PartialTopMenu
checkAccount conditions = undefined -- Implementation goes here
